src/utilities.jl

```diff
@@ -32,7 +32,7 @@ function methodgroups(func, typesig, modname; exact = true)
     for (key, group) in groups
         filter!(group) do m
             local ismod = m.module == modname
-            exact ? (ismod && Base.tuple_type_tail(m.sig) in typesigs) : ismod
+            exact ? (ismod && Base.rewrap_unionall(Base.tuple_type_tail(m.sig), m.sig) in typesigs) : ismod
```

Comment:
To clarify: for `UnionAll` types, `Base.tuple_type_tail` calls the 'regular old' `Base.tuple_type_tail` on the `body` field of the `UnionAll` type; see [here](https://github.com/JuliaLang/julia/blob/903644385b91ed8d95e5e3a5716c089dd1f1b08a/base/essentials.jl#L68). Getting the `body` field is exactly what [`unwrap_unionall`](https://github.com/JuliaLang/julia/blob/903644385b91ed8d95e5e3a5716c089dd1f1b08a/base/essentials.jl#L84) does. We need to call [`rewrap_unionall`](https://github.com/JuliaLang/julia/blob/903644385b91ed8d95e5e3a5716c089dd1f1b08a/base/essentials.jl#L91) to put the free type variables back in place after performing an operation on just the `body` field.

Here's an example:
```julia
julia> foo = Tuple{identity,Union{Array{T,1}, Array{T,2}}} where T
Tuple{identity,Union{Array{T,1}, Array{T,2}}} where T

julia> typeof(foo)
UnionAll

julia> foo.body
Tuple{identity,Union{Array{T,1}, Array{T,2}}}

julia> typeof(foo.body)
DataType

julia> tail = Base.tuple_type_tail(foo)
Tuple{Union{Array{T,1}, Array{T,2}}}

julia> typeof(tail)
DataType

julia> Base.rewrap_unionall(tail, foo)
Tuple{Union{Array{T,1}, Array{T,2}}} where T
```